### PR TITLE
Added an option for strict neighbor list

### PR DESCRIPTION
**Pull request by ceriottm**
Adds a flag to the neighbor list classes to signal whether the neighbor list is "strict" - that is if it guarantees not to pass atoms beyond the cutoff. 

# Contributor (creator of pull-request) checklist

 - [ ] Tests updated (for new features and bugfixes)?
 - [ ] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/metatensor/metatensor/actions/artifacts/2116989539.zip)

### CODE REVIEW

File: python/examples/atomistic/3-atomistic-model-with-nl.py [hunk 1]
```diff
@@ -163,11 +163,12 @@
 # %%
 #
 # The data and metadata inside the ``neighbors`` object do not contain information about
-# the ``cutoff`` and whether this is a full or half neighbor list. To account for this,
+# the ``cutoff``, whether this is a full or half neighbor list, and whether it is
+# restricted to distances strictly below the cutoff (strict). To account for this,
```
Comment (Luthaf) [inline python/examples/atomistic/3-atomistic-model-with-nl.py:167]:
```suggestion
# restricted to distances strictly below the cutoff (``strict=True``). To account for this,
```

File: metatensor-torch/tests/neighbor-checks/slanted-loose.json [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        "cell": [
+            [4.26, -2.45951215, 0.0],
+            [2.13, 1.22975607, 0.0],
+            [0.0, 0.0, 50.0]
+        ]
+    },
+    "options": {
+        "cutoff": 4.1,
+        "full_list": false,
+        "strict": false
```
Comment (Luthaf) [inline metatensor-torch/tests/neighbor-checks/slanted-loose.json:19]: I'm not sure what this is intended to test? The set of pairs included in a non-strict NL will depend a lot on the actual implementation. I think these files are more about checking the strict case, and in the non-strict case the set of `expected-neighbors` is the minimal set of pairs that should be included.

File: python/examples/atomistic/3-atomistic-model-with-nl.py [hunk 3]
```diff
@@ -163,11 +163,12 @@
 # %%
 #
 # The data and metadata inside the ``neighbors`` object do not contain information about
-# the ``cutoff`` and whether this is a full or half neighbor list. To account for this,
-# metatensor neighbor lists are always stored together with
+# the ``cutoff``, whether this is a full or half neighbor list, and whether it is
+# restricted to distances strictly below the cutoff (``strict=True``). To account
+# for this, metatensor neighbor lists are always stored together with
```
Comment (Luthaf) [inline python/examples/atomistic/3-atomistic-model-with-nl.py:168]: We should have a small explanation about when someone would want to use non-strict list & how to properly hand them.

File: .github/workflows/python-tests.yml [hunk 4]
```diff
@@ -22,7 +22,7 @@ jobs:
             torch-version: "1.12.*"
             numpy-version-pin: "<2.0"
           - os: ubuntu-20.04
-            python-version: "3.9"
+            python-version: "3.10"
```
Comment (Luthaf) [inline .github/workflows/python-tests.yml:25]: @ceriottm could you please not force push back changes I undid?

**Review by Luthaf — APPROVED** [review]
This should be good to go!

File: .github/workflows/python-tests.yml (hunk 4 above)
Comment (ceriottm) [reply .github/workflows/python-tests.yml]: Don't know how this happened TBH. ohshitgit I guess

**Pull request merged**